### DISCUSSION TITLE
Testing martin with MapLibre from a pbf file

### DISCUSSION THREAD

**Discussion by BigBoulard** [maplibre/martin#1233] [Q&A] [answered]
Hi guys,

My goal is to replace Google Maps with a self-hosted solution. I'm a total newbie in GIS so it's a bit challenging ...
TLDR: I don't know if the bug comes from the MapLibre code or Martin. I just created a repo for testing, it tries to display a MapLibre map of Brussels City based on the Belgium `mbtiles` from `https://download.geofabrik.de/europe/belgium-latest.osm.pbf`

<https://github.com/BigBoulard/martin-test>

Here is what I did so far:

## 1. Frontend Test Page with a minimal  `osm-style.json`

I started from an existing style.json and changed the reference to my tile server (Martin):

```json
"sources": {
    "osm-tiles": {
      "type": "vector",
      "tiles": [
        "https://mygateway:8080/planet_osm_point,planet_osm_line,planet_osm_roads,planet_osm_polygon"
      ],
      "minzoom": 0,
      "maxzoom": 22
    }
  },
```

 then I changed the `source` and `source-layer` fields in each layer trying to guess if the data is a `point`, `line`, `polygon`, or a `road` and reference the source (i`planet_osm_polygon` below):

```json
"source": "osm-tiles",
"source-layer": "planet_osm_polygon",
```

```js
var map = new maplibregl.Map({
	container: 'map',
	style: '/osm-style.json',
	center: [4.3601755242029885, 50.85819161299564], // brussels
	zoom: 13 // starting zoom
});
```

**osm-style.json**
```json
{
  "version": 8,
  "sources": {
    "osm-tiles": {
      "type": "vector",
      "tiles": [
         "https://mygateway:8080/planet_osm_point,planet_osm_line,planet_osm_roads,planet_osm_polygon"
      ],
      "minzoom": 0,
      "maxzoom": 22
    }
  },
  "layers": [
    {
      "id": "background",
      "type": "background",
      "paint": {
        "background-color": "#f8f4f0"
      }
    },
    {
      "id": "water",
      "type": "fill",
      "source": "osm-tiles",
      "source-layer": "planet_osm_polygon",
      "filter": ["all", ["!=", "intermittent", 1], ["!=", "brunnel", "tunnel"]],
      "layout": {"visibility": "visible"},
      "paint": {"fill-color": "hsl(210, 67%, 85%)"}
    }    
  ]
}
```

## 3. Import of the `belgium-latest.osm.pbf` file into postgres

```sh
osm2pgsql -c -d osm -U osm_user -H localhost -P 5432 -W -C 2000 belgium-latest.osm.pbf
```

## 4. Ran `Martin Tile Server` in debug mod and check that tiles can be served

When I hit `http://0.0.0.0:3000/catalog`, I see that Martin finds `postgres` tables, sprites and fonts.
```json
{
  "tiles": {
    "planet_osm_line": {
      "content_type": "application/x-protobuf",
      "description": "public.planet_osm_line.way"
    },
    "planet_osm_point": {
      "content_type": "application/x-protobuf",
      "description": "public.planet_osm_point.way"
    },
    "planet_osm_polygon": {
      "content_type": "application/x-protobuf",
      "description": "public.planet_osm_polygon.way"
    },
    "planet_osm_roads": {
      "content_type": "application/x-protobuf",
      "description": "public.planet_osm_roads.way"
    }
  },
  "sprites": {
    "my_sprites": {
      "images": [...]
    }
  },
  "fonts": {...}
}
```

http://0.0.0.0:3000/planet_osm_point,planet_osm_line,planet_osm_roads,planet_osm_polygon
```json
{
  "tilejson": "3.0.0",
  "tiles": [
    "http://0.0.0.0:3000/planet_osm_point,planet_osm_line,planet_osm_roads,planet_osm_polygon/{z}/{x}/{y}"
  ],
  "vector_layers": [
    {
      "id": "planet_osm_point",
      "fields": {} // 68 items
    },
    {
      "id": "planet_osm_line",
      "fields": {} // 68 items
    },
    {
      "id": "planet_osm_roads",
      "fields": {} // 68 items
    },
    {
      "id": "planet_osm_polygon",
      "fields": {} // 68 items
    }
  ],
  "bounds": [
    -5.652016999999999, 49.41460299959042, 7.2824670000000005, 58.36698750096329
  ],
  "description": "public.planet_osm_point.way\npublic.planet_osm_line.way\npublic.planet_osm_roads.way\npublic.planet_osm_polygon.way",
  "name": "planet_osm_point,planet_osm_line,planet_osm_roads,planet_osm_polygon"
}
```
## Problem

Here is what I get when the test page loads:
![image](https://github.com/maplibre/martin/assets/60930059/4a419d9e-913f-4dfe-8f78-434f9ef9e0ca)

There's no shape and I'm not supposed to see that much water around Brussels.

### Martin Logs upon page load

```sh
2024-03-07 11:35:48 [2024-03-07T10:35:48Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_roads/13/4196/2747 HTTP/1.1" 200 13889 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.111819
2024-03-07 11:35:48 [2024-03-07T10:35:48Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_point/13/4194/2747 HTTP/1.1" 200 11677 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.120157
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_roads/13/4195/2748 HTTP/1.1" 200 25702 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.159577
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_line/13/4196/2748 HTTP/1.1" 200 24327 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.168348
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_line/13/4194/2746 HTTP/1.1" 200 25263 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.174563
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4195/2747 HTTP/1.1" 200 26338 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.177000
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4194/2748 HTTP/1.1" 200 27261 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.178879
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4196/2746 HTTP/1.1" 200 31472 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.190088
2024-03-07 11:35:49 [2024-03-07T10:35:49Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_point/13/4195/2746 HTTP/1.1" 200 10210 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.050309
```
### Martin Logs with a sprite reference in style.json and some more layers

```sh
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse clip-path value: 'none'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse clip-path value: 'none'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse clip-path value: 'none'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse clip-path value: 'none'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse stroke-miterlimit value: '4px'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z WARN  usvg_parser::svgtree] Failed to parse clip-path value: 'none'.
2024-03-07 16:28:40 [2024-03-07T15:28:40Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /sprite/my_sprite.json HTTP/1.1" 200 1121 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.086412
2024-03-07 16:28:40 [2024-03-07T15:28:40Z INFO  oxipng] Processing from memory
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     256x141 pixels, PNG format
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     8-bit RGB + Alpha, non-interlaced
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     IDAT size = 43789 bytes
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     File size = 43846 bytes
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng] Trying: None
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng] Found better combination:
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     zc = 11  f = None      11038 bytes
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     IDAT size = 11038 bytes (32751 bytes decrease)
2024-03-07 16:28:40 [2024-03-07T15:28:40Z DEBUG oxipng]     file size = 11095 bytes (32751 bytes = 74.70% decrease)
2024-03-07 16:28:40 [2024-03-07T15:28:40Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /sprite/my_sprite.png HTTP/1.1" 200 11095 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.151146
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_point/13/4194/2747 HTTP/1.1" 200 11677 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.045285
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_roads/13/4196/2747 HTTP/1.1" 200 13889 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.063490
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_point/13/4195/2746 HTTP/1.1" 200 10210 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.071279
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4195/2747 HTTP/1.1" 200 26420 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.121468
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_line/13/4194/2746 HTTP/1.1" 200 23948 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.111331
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_roads/13/4195/2748 HTTP/1.1" 200 25702 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.112891
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4196/2746 HTTP/1.1" 200 31472 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.131062
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_polygon/13/4194/2748 HTTP/1.1" 200 27261 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.107893
2024-03-07 16:28:41 [2024-03-07T15:28:41Z INFO  actix_web::middleware::logger] 192.168.65.1 "GET /planet_osm_line/13/4196/2748 HTTP/1.1" 200 24327 "https://localhost:5173/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" 0.075106
```

Many thanks

**Comment by BigBoulard** [comment, marked as answer]
There were several issues.

1. The best method to test self-served tiles with Martin is to generate the tiles with `Planetiler` which produces a `mbtiles` file that Martin can use as a source to serve.

```sh
docker run -e JAVA_TOOL_OPTIONS="-Xmx1g" -v "$(pwd)/planetiler":/data ghcr.io/onthegomap/planetiler:latest --download --area=belgium
```
see [planetiler](https://github.com/onthegomap/planetiler/blob/main/README.md)

```yaml
version: '3.8'
services:
  martin:
    image: ghcr.io/maplibre/martin:v0.13.0
    ports:
      - 3000:3000
    environment:
      - RUST_LOG=debug
    volumes:
      - ./fonts:/fonts
      - ./icons:/sprite
      - ./maps-data/belgium.mbtiles:/maps-data/belgium.mbtiles
    command: /maps-data/belgium.mbtiles --sprite /sprite --font /fonts
```

2. A JSON endpoint must be referenced into the style, and there are several possibilities, but the easiest one is to use a single source URL referencing the `mbtile` file like this: 
```json
   "osmtiles": {      
      "type": "vector",
      "tiles": ["http://0.0.0.0:3000/belgium/{z}/{x}/{y}"],
      "minzoom": 0,
      "maxzoom": 22
   }
```

Now the map displays smoothly. I just noticed in the JS Console that if I zoom in too much, I get warnings about missing icons in the sprite, this is [a known issue in the osm bright style](https://github.com/openmaptiles/osm-bright-gl-style/issues/47)

```sh
...
Image "bicycle_parking_11" could not be loaded. Please make sure you have added the image with map.addImage() or a "sprite" property in your style. You can provide missing images by listening for the "styleimagemissing" map event.

Image "doctors_11" could not be loaded. Please make sure you have added the image with map.addImage() or a "sprite" property in your style. You can provide missing images by listening for the "styleimagemissing" map event.
...
```
